File: COURS/MLB/LABS/Ozone_LinearModel_1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Apr 11 13:48:37 2019

@author: valerie
"""

import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import linear_model
from urllib.request import urlopen


# Data importation
# On lit le tableau comme un tableau de chaînes de caractères,
# on extrait les colonnes numériques, 
# on fabrique artisanalement les dummies avec les colonnes catégorielles,
# on reconstitue le tout.
url="https://perso.univ-rennes1.fr/bernard.delyon/tp/ozone.txt"
Ozone = np.loadtxt(urlopen(url),skiprows=1,dtype=np.ndarray)
nomvar=np.genfromtxt(urlopen(url),max_rows=1,dtype='str')
y=Ozone[:,[1]].astype(float)
X=Ozone[:,2:-2]
vent=Ozone[:,[-2]]
VO=(vent=='Ouest')*1   # categorical variables have to be recoded
VS=(vent=='Sud')*1
VE=(vent=='Est')*1
P=(Ozone[:,[-1]]=='Pluie')*1
X=np.concatenate((X,VO,VE,VS,P),axis=1).astype(float)
# Noms des variables explicatives
nomvar=np.concatenate((nomvar[1:-2],["Ouest"],["Sud"],["Est"],["Pluie"]))



lr = linear_model.LinearRegression(fit_intercept=True)

B = 30 # number of samples for the cross validation
mse_lr, mse_pca = np.zeros(B),np.zeros(B)
for b in range(0,B):
    # sampling for train, test sets
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25)

    # standardization
    numvar = np.arange(10) # list of quantitative variables
    mk=np.mean(X_train[:,numvar],axis=0)
    sk=np.maximum(np.std(X_train[:,numvar],axis=0),10*np.finfo(float).eps)
    X_train[:,numvar], X_test[:,numvar] = np.add(X_train[:,numvar],-mk), np.add(X_test[:,numvar],-mk)
    X_train[:,numvar], X_test[:,numvar] = np.multiply(X_train[:,numvar],1/sk),np.multiply(X_test[:,numvar],1/sk)
    
    # learning linear model
    lr.fit(X_train,y_train) # fit the linear model
    y_pred = lr.predict(X_test) # prediction for the test set
    mse_lr[b] = np.mean((y_test-y_pred)**2) # compute MSE
    
    y_pred_naive = np.mean(y_train)
    mse_naive = np.mean((y_test-y_pred_naive)**2)
    
    

print("Error of linear regression: ", str(np.round(np.sqrt(np.mean(mse_lr)),2)))
print("Error of naive prediction: ", str(np.round(np.sqrt(np.mean(mse_naive)),2)))

File: COURS/MLB/LABS/Ozone_LinearModel_1.R


# Data importation
# On lit le tableau comme un tableau de chaînes de caractères,
# on extrait les colonnes numériques, 
# on fabrique artisanalement les dummies avec les colonnes catégorielles,
# on reconstitue le tout.
url="https://perso.univ-rennes1.fr/bernard.delyon/tp/ozone.txt"
Ozone = read.table(url,header=TRUE)

y=Ozone[,1]
X=Ozone[,-1]


train_test_split <- function (X,y,test_size=.25,random_state=NULL){
# Extraction of a train and a test datasets.
#
# Inputs : X, y : data
#          test_size : a fraction (<1) of the total set or a number of samples (integer) ; default 0.25
#          random_state : if equal to an interger, it fixes the random seed ; defaut NULL
# Outputs : X_train, X_test, y_train, y_test
#
  n = nrow(X)
  if (test_size>1){test_size=test_size/n}
  if (!is.null(random_state)){set.seed(random_state)}
  itest=sample(1:n,round(n*test_size))
  itrain=setdiff(1:n,itest)
  Xtrain=X[itrain,]
  Xtest=X[itest,]
  ytrain=y[itrain]
  ytest=y[itest]
  return(list(X_train=Xtrain,X_test=Xtest,y_train=ytrain,y_test=ytest))
}

B = 30 # number of samples for the cross validation
mse_lr = mse_naive = matrix(0,B,1)
for (b in 1:B){
  # sampling for train, test sets
  samples = train_test_split(X,y,test_size=0.25,random_state=1)
  X_train=samples$X_train
  X_test=samples$X_test
  y_train=samples$y_train
  y_test=samples$y_test

  # standardization
  numvar = 1:10 # list of quantitative variables
  mk=apply(X_train[,numvar],2,mean)
  sk=apply(X_train[,numvar],2,sd)
  X_train[,numvar] = scale(X_train[,numvar],center=mk,scale=sk)
  X_test[,numvar] = scale(X_test[,numvar],center=mk,scale=sk)
  data_train = data.frame(X_train,y_train)
  data_test = data.frame(X_test,y_test)
  
  # learning linear model
  lr = lm(y_train~.,data=data_train)
  y_pred = predict(lr,data_test) # prediction for the test set
  mse_lr[b] = mean((y_test-y_pred)^2) # compute MSE

  y_pred_naive = mean(y_train)
  mse_naive = mean((y_test-y_pred_naive)^2)
}

print(paste("Error of linear regression: ", round(sqrt(mean(mse_lr))*100)/100))
print(paste("Error of naive prediction: ", round(sqrt(mean(mse_naive))*100)/100))


